lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,383 @@
+  DenseSet<Instruction *> canAssumeUniforms = std::move(initialReadFirstLanes);
+
+  for (BasicBlock::reverse_iterator instIt = block->rbegin(), endIt = block->rend(); instIt != endIt; ++instIt) {
+    Instruction *inst = &*instIt;
+    // Skip readfirstlane-irrelevant instructions
+    if (canAssumeUniforms.count(inst) == 0)
+      continue;
+    // A value can be "assumed uniform" should meet the following conditions:
+    // 1. The value itself is divergent instruction
+    // 2. Tha value itself and all its users are in the testing basic block
+    // 3. All users of the value are "assumed uniforms"
+    for (Use &use : inst->operands()) {
+      if (!m_divergenceAnalysis->isDivergentUse(&use))
+        continue;
+      auto operandInst = dyn_cast<Instruction>(use.get());
+      if (!operandInst)
+        continue;
+      if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block))
+        break;
+      if (!isAllUsersAssumedUniform(operandInst, canAssumeUniforms))
+        continue;
+      m_canAssumeUniformDivergentUsesMap[inst].insert(operandInst);
+      if (canAssumeUniforms.count(operandInst) == 0 && !m_targetTransformInfo->isSourceOfDivergence(operandInst)) {
+        m_canAssumeUniformDivergentUsesMap.insert({operandInst, {}});
+        canAssumeUniforms.insert(operandInst);
+      }
+    }
+  }
+}
+
+// =====================================================================================================================
+// Find the best insert locations according to the heuristic.
+// The heuristic is: if an instruction that can be assumed to be uniform has multiple divergent operands, then you take
+// the definition of the divergent operand that is earliest in basic block order (call it "the earliest divergent
+// operand") and propagate up to that instruction; if it turns out that that instruction can be assumed to be uniform,
+// then we can just insert the readfirstlane there (or propagate).
+//
+// @param block : The process basic block
+void PatchReadFirstLane::findInsertLocations(BasicBlock *block) {
+  m_insertLocations.clear();
+  // Find insert locations or propagate up among "assumed uniforms" by backward traversing the basic block
+  for (auto assumedUniformDivergentUses : m_canAssumeUniformDivergentUsesMap) {
+    Instruction *candidateInst = assumedUniformDivergentUses.first;
+
+    auto &divergentDefSet = assumedUniformDivergentUses.second;
+    const unsigned divergentDefCount = divergentDefSet.size();
+    if (divergentDefCount == 0) {
+      // No divergent use for propagating up and try the candidateInst as an insert location
+      addLocation(candidateInst);
+      continue;
+    }
+    Instruction *earliestDivergentDef = *divergentDefSet.begin();
+    if (divergentDefCount == 1 && m_canAssumeUniformDivergentUsesMap.count(earliestDivergentDef) == 0) {
+      earliestDivergentDef = nullptr;
```

Comment:
Can this special case for divergentDefCount == 1 somehow be folded into the general case?